app/models/search_facet.rb

```diff
@@ -0,0 +1,165 @@
+    }
+  end
+
+  # retrieve unique values from BigQuery and format an array of hashes with :name and :id values to populate :filters attribute
+  def get_unique_filter_values
+    Rails.logger.info "Updating filter values for SearchFacet: #{self.name} using id: #{self.big_query_id_column} and name: #{self.big_query_name_column}"
+    queries = []
+    if self.is_array_based
+      queries << self.generate_array_query(self.big_query_id_column, 'id')
+      queries << self.generate_array_query(self.big_query_name_column, 'name')
+    else
+      queries << self.generate_non_array_query
+    end
+    begin
+      results = []
+      queries.each do |query_string|
+        Rails.logger.info "Executing query: #{query_string}"
+        results << SearchFacet.big_query_dataset.query(query_string)
+      end
+      return assemble_filters_array(results)
+    rescue => e
+      Rails.logger.error "Error retrieving unique values for #{CellMetadatum::BIGQUERY_TABLE}: #{e.class.name}:#{e.message}"
+      error_context = ErrorTracker.format_extra_context({queries: queries})
+      ErrorTracker.report_exception(e, nil, error_context)
+      []
+    end
+  end
+
+  # update cached filters in place with new values
+  def update_filter_values!
+    values = self.get_unique_filter_values.to_a
+    unless values.empty?
+      self.update(filters: values)
+    else
+      false # did not get any results back, meaning :retrieve_unique_filter_values encountered an error
+    end
+  end
+
+  # generate a single query to get DISTINCT values from an array-based column, preserving order
+  # this way we can do two queries, and stitch them together in a hash as the order will be the same
+  # e.g. IDs will line up with NAMEs
+  def generate_array_query(column_name, identifier)
+    "SELECT DISTINCT #{identifier} FROM(SELECT array_col AS #{identifier}, " + \
+    "FROM #{CellMetadatum::BIGQUERY_TABLE}, UNNEST(#{column_name}) AS array_col " + \
+    "WITH OFFSET AS offset ORDER BY offset)"
+  end
+
+  # generate query string to retrieve distinct values for non-array based facets
+  def generate_non_array_query
+    "SELECT DISTINCT #{self.big_query_id_column} AS id, #{self.big_query_name_column} AS name FROM #{CellMetadatum::BIGQUERY_TABLE}"
+  end
+
+  # stitch together results into the formatted filters array based on whether or not we ran one or two queries
+  def assemble_filters_array(results_array)
+    if results_array.size == 1
+      results_array.first # already in the correct form, so just return
+    else
+      # we have two hashes we need to merge together
+      filters_array = []
+      id_array = results_array.first
+      name_array = results_array.last
+      id_array.each_with_index do |filter_hash, index|
+        filters_array << filter_hash.merge(name_array[index])
+      end
+      filters_array
+    end
+  end
+
+  private
+
+  # determine if this facet references array-based data in BQ as data_type will look like "ARRAY<STRING>"
```

Comment:
maybe rename to 'set_is_array_based_from_bq'?

Reply:
That makes sense.

Reply:
4d685ac

Reply:
I think you forgot to make the change